### PR TITLE
chore(compiler): type references as expressions and phase tracking

### PR DESCRIPTION
**Pull request by eladb**
No functional changes, preparatory pull request as part of the expression lifting escapade.

Type references are expressions that reference type names. It's valuable to model them as expressions because they are needed at runtime. Once they are modeled as expressions, we can treat all expressions equally when it comes to lifting (which is the motivation for this refactor).

We've changed the following places to use type references instead of a `UserDefinedType`. In all of these cases the types have a runtime presence. The rest of the places are just type annotations that are erased at runtime:

- The class being instantiated in `new` expressions
- Base classes
- The type component in a `Reference::TypeMember` (static access).

Furthermore, to support expression lifting, the type checker now records the phase of each expression during type checking and makes it available through `get_expr_phase`.

Additionally, changed the left-hand-side of an assignment statement to be an `Expr` and not a `Reference`. Also in order to be able to lift it.

Turn the variable component of `StmtKind::Assignment` from a simple `Symbol` to an `Expr` in order for it to also hold type and phase information.

The JSII importer accidentally imported structs as `preflight` while they should be phase-independent.

Misc: strip excess newlines when adding a line to `CodeMaker`.

## Checklist

- [x] Title matches [Winglang's style guide](https://docs.winglang.io/contributing/pull_requests#how-are-pull-request-titles-formatted)
- [x] Description explains motivation and solution
- [x] Tests added (always)
- [x] Docs updated (only required for features)
- [x] Added `pr/e2e-full` label if this feature requires end-to-end testing

*By submitting this pull request, I confirm that my contribution is made under the terms of the [Monada Contribution License](https://docs.winglang.io/terms-and-policies/contribution-license.html)*.

### CODE REVIEW

File: tools/hangar/__snapshots__/invalid.ts.snap [hunk 1]
```diff
@@ -44,6 +44,20 @@ error: Unknown symbol \\"this\\"
   |     ^^^^ Unknown symbol \\"this\\"
 
 
+error: Unknown symbol \\"this\\"
```
Comment (staycoolcall911) [inline tools/hangar/__snapshots__/invalid.ts.snap:47]: What introduced this change in this snapshot?

**Review by staycoolcall911 — APPROVED** [review]
Thanks to the detailed PR description reviewing was a breeze...

File: tools/hangar/__snapshots__/invalid.ts.snap (hunk 1 above)
Comment (eladb) [reply tools/hangar/__snapshots__/invalid.ts.snap:47]: Checking

File: tools/hangar/__snapshots__/invalid.ts.snap [hunk 2]
```diff
@@ -280,11 +279,11 @@ error: Preflight field \\"y\\" is not initialized
    |   ^ Preflight field \\"y\\" is not initialized
 
 
-error: Expected \\"x\\" to be a type but it's a variable
+error: Base class \\"unresolved\\" is not a class
    --> ../../../examples/tests/invalid/class.w:68:18
    |
 68 | class C7 extends x {
-   |                  ^ Expected \\"x\\" to be a type but it's a variable
+   |                  ^ Base class \\"unresolved\\" is not a class
```
Comment (MarkMcCulloh) [inline tools/hangar/__snapshots__/invalid.ts.snap:286]: Regression

File: tools/hangar/__snapshots__/invalid.ts.snap [hunk 3]
```diff
@@ -294,11 +293,11 @@ error: Base class \\"S1\\" is not a class
    |                  ^^ Base class \\"S1\\" is not a class
 
 
-error: Class cannot extend itself
+error: Base class \\"unresolved\\" is not a class
    --> ../../../examples/tests/invalid/class.w:78:19
    |
 78 | class C11 extends C11 {
-   |                   ^^^ Class cannot extend itself
+   |                   ^^^ Base class \\"unresolved\\" is not a class
```
Comment (MarkMcCulloh) [inline tools/hangar/__snapshots__/invalid.ts.snap:300]: Regression

File: tools/hangar/__snapshots__/invalid.ts.snap [hunk 4]
```diff
@@ -336,6 +335,13 @@ error: Expected type to be \\"num\\", but got \\"str\\" instead
    |              ^^^^ Expected type to be \\"num\\", but got \\"str\\" instead
 
 
+error: Variable cannot be reassigned from inflight
+   --> ../../../examples/tests/invalid/class.w:61:5
+   |
+61 |     this.y = 1;
+   |     ^^^^^^^^^^^ Variable cannot be reassigned from inflight
```
Comment (MarkMcCulloh) [inline tools/hangar/__snapshots__/invalid.ts.snap:342]: This line already has a more specific error above:

```
"y" cannot be initialized in the inflight initializer
```

Technically y is not even reassignable at all, so `cannot be reassigned from inflight` is true but misleading.

File: tools/hangar/__snapshots__/invalid.ts.snap [hunk 5]
```diff
@@ -1965,11 +1964,11 @@ Duration <DURATION>"
...
 
 exports[`void_in_expression_position.w 1`] = `
-"error: Property access unsupported on type \\"void\\"
-  --> ../../../examples/tests/invalid/void_in_expression_position.w:1:1
+"error: Property not found
+  --> ../../../examples/tests/invalid/void_in_expression_position.w:1:12
   |
 1 | log(\\"hey\\").get(\\"x\\");
-  | ^^^^^^^^^^ Property access unsupported on type \\"void\\"
+  |            ^^^ Property not found
```
Comment (MarkMcCulloh) [inline tools/hangar/__snapshots__/invalid.ts.snap:1971]: This diagnostic is accurate but misleading. The original diagnostic does need some work though so it's probably fine to try this new one out.

File: tools/hangar/__snapshots__/invalid.ts.snap (hunk 2 above)
Comment (eladb) [reply tools/hangar/__snapshots__/invalid.ts.snap:286]: yes, I am on it!

File: tools/hangar/__snapshots__/invalid.ts.snap (hunk 5 above)
Comment (eladb) [reply tools/hangar/__snapshots__/invalid.ts.snap:1971]: Yes, that was my thought :-)

File: tools/hangar/__snapshots__/invalid.ts.snap (hunk 4 above)
Comment (eladb) [reply tools/hangar/__snapshots__/invalid.ts.snap:342]: Good point. Looking into this.

File: tools/hangar/__snapshots__/invalid.ts.snap [hunk 6]
```diff
@@ -1528,11 +1527,11 @@ exports[`sorted_errors_no_span.w 1`] = `
   |              ^^^ Expected type to be \\"num\\", but got \\"str\\" instead
 
 
-error: Expected \\"b\\" to be a type but it's a variable
+error: Base class \\"unresolved\\" is not a class
   --> ../../../examples/tests/invalid/sorted_errors_no_span.w:3:26
   |
 3 | inflight class c extends b {}
-  |                          ^ Expected \\"b\\" to be a type but it's a variable
+  |                          ^ Base class \\"unresolved\\" is not a class
```
Comment (eladb) [inline tools/hangar/__snapshots__/invalid.ts.snap:1534]: yak

File: tools/hangar/__snapshots__/invalid.ts.snap [hunk 7]
```diff
@@ -1864,32 +1863,32 @@ Duration <DURATION>"
 `;
 
 exports[`unknown_symbol.w 1`] = `
-"error: Unknown symbol \\"clod\\"
-  --> ../../../examples/tests/invalid/unknown_symbol.w:3:18
```
Comment (eladb) [inline tools/hangar/__snapshots__/invalid.ts.snap:1868]: regression

File: tools/hangar/__snapshots__/invalid.ts.snap [hunk 8]
```diff
@@ -1864,32 +1863,32 @@ Duration <DURATION>"
...
-error: Unknown symbol \\"y\\"
+"error: Unknown symbol \\"y\\"
   --> ../../../examples/tests/invalid/unknown_symbol.w:9:13
   |
 9 | let x = 2 + y;
   |             ^ Unknown symbol \\"y\\"
 
 
-error: Unknown symbol \\"B\\"
+error: Base class \\"unresolved\\" is not a class
```
Comment (eladb) [inline tools/hangar/__snapshots__/invalid.ts.snap:1873]: regression

File: tools/hangar/__snapshots__/invalid.ts.snap (hunk 1 above)
Comment (eladb) [reply tools/hangar/__snapshots__/invalid.ts.snap:47]: This is actually the correct errors.

File: tools/hangar/__snapshots__/invalid.ts.snap [hunk 9]
```diff
@@ -280,11 +294,18 @@ error: Preflight field \\"y\\" is not initialized
    |   ^ Preflight field \\"y\\" is not initialized
 
 
-error: Expected \\"x\\" to be a type but it's a variable
+error: Unable to resolve type \\"x\\"
    --> ../../../examples/tests/invalid/class.w:68:18
    |
 68 | class C7 extends x {
-   |                  ^ Expected \\"x\\" to be a type but it's a variable
```
Comment (eladb) [inline tools/hangar/__snapshots__/invalid.ts.snap:287]: regression

File: tools/hangar/__snapshots__/invalid.ts.snap [hunk 10]
```diff
@@ -280,11 +294,18 @@ error: Preflight field \\"y\\" is not initialized
...
+   |                  ^ Unable to resolve type \\"x\\"
+
+
+error: Base class \\"unresolved\\" is not a class
+   --> ../../../examples/tests/invalid/class.w:68:18
+   |
+68 | class C7 extends x {
+   |                  ^ Base class \\"unresolved\\" is not a class
```
Comment (eladb) [inline tools/hangar/__snapshots__/invalid.ts.snap:308]: regression

File: tools/hangar/__snapshots__/invalid.ts.snap [hunk 11]
```diff
@@ -514,18 +549,11 @@ Duration <DURATION>"
 `;
 
 exports[`enums.w 1`] = `
-"error: Enum \\"SomeEnum\\" does not contain value \\"FOUR\\"
-  --> ../../../examples/tests/invalid/enums.w:5:21
-  |
-5 | let four = SomeEnum.FOUR;
-  |                     ^^^^ Enum \\"SomeEnum\\" does not contain value \\"FOUR\\"
```
Comment (eladb) [inline tools/hangar/__snapshots__/invalid.ts.snap:521]: Regression

File: libs/wingc/src/type_check.rs [hunk 12]
```diff
@@ -2355,12 +2388,23 @@ impl<'a> TypeChecker<'a> {
...
-					self.spanned_error(stmt, format!("Variable {} is not reassignable ", variable));
+				let (exp_type, _) = self.type_check_exp(value, env);
+				let (var_type, var_phase) = self.type_check_exp(variable, env);
+
+				// check if the variable can be reassigned
+
+				if let ExprKind::Reference(r) = &variable.kind {
+					let (var, _) = self.resolve_reference(&r, env);
+					if var_phase == Phase::Preflight && env.phase == Phase::Inflight {
+						self.spanned_error(stmt, "Variable cannot be reassigned from inflight".to_string());
```
[the remark this answers is not in the document]
Comment (eladb) [reply libs/wingc/src/type_check.rs:2439]: added

File: tools/hangar/__snapshots__/invalid.ts.snap (hunk 4 above)
Comment (eladb) [reply tools/hangar/__snapshots__/invalid.ts.snap:371]: Fixed.

**Comment by eladb** [comment]
@MarkMcCulloh I'll make sure all the diagnostics around base classes are fixed. If that's all, I'd appreciate an approval so I can be unblocked with my next step.

**Review by MarkMcCulloh — APPROVED** [review]
> I'll make sure all the diagnostics around base classes are fixed. If that's all, I'd appreciate an approval so I can be unblocked with my next step.

Yeah LGTM as long as the diags are fixed. Luckily the grammar still prevents real expressions into places where types are expected so we might be safe from some new big regressions.

I also think @yoav-steinberg should take a look at this, but I'll leave that to you if you want to wait.

File: libs/wingc/src/type_check.rs [hunk 13]
```diff
@@ -74,21 +74,40 @@ pub enum SymbolKind {
 	Namespace(NamespaceRef),
 }
 
+#[derive(Debug, Clone)]
+pub enum VariableKind {
+	/// a free variable not associated with a specific type
+	Free,
```
Comment (Chriscbr) [inline libs/wingc/src/type_check.rs:80]: Is this meant to refer to any local variable? I think that might be better compiler nomenclature:
```suggestion
	/// a local variable not associated with a specific type
	Local,
```

File: libs/wingc/src/type_check.rs [hunk 14]
```diff
@@ -1144,18 +1169,28 @@ impl Types {
 	}
 
 	/// Stores the type of a given expression node.
```
Comment (Chriscbr) [inline libs/wingc/src/type_check.rs:1171]:
```suggestion
	/// Stores the type and phase of a given expression node.
```

File: libs/wingc/src/type_check.rs [hunk 15]
```diff
@@ -1244,61 +1288,61 @@ impl<'a> TypeChecker<'a> {
...
+		self.types.assign_type_to_expr(exp, t, phase);
+		(t, phase)
 	}
 
 	/// Helper function for type_check_exp. This is needed because we want to be able to `return`
 	/// and break early, while still setting the evaluated type on the expression.
 	///
 	/// Do not use this function directly, use `type_check_exp` instead.
-	fn type_check_exp_helper(&mut self, exp: &Expr, env: &SymbolEnv) -> TypeRef {
+	fn type_check_exp_helper(&mut self, exp: &Expr, env: &SymbolEnv) -> (TypeRef, Phase) {
```
Comment (Chriscbr) [inline libs/wingc/src/type_check.rs:1313]: This is pretty cool!

File: libs/wingc/src/type_check.rs [hunk 16]
```diff
@@ -1244,61 +1288,61 @@ impl<'a> TypeChecker<'a> {
...
 					s.parts.iter().for_each(|part| {
 						if let InterpolatedStringPart::Expr(interpolated_expr) = part {
-							let exp_type = self.type_check_exp(interpolated_expr, env);
+							let (exp_type, _) = self.type_check_exp(interpolated_expr, env);
 							self.validate_type_in(exp_type, &self.types.stringables(), interpolated_expr);
 						}
 					});
-					self.types.string()
+					// TODO: do we need to determine the phase based on the inner expressions?
+					(self.types.string(), Phase::Independent)
```
Comment (Chriscbr) [inline libs/wingc/src/type_check.rs:1326]: I think as long as strings are phase-independent, it seems okay to me to say that the phase of the template string is independent

File: libs/wingc/src/type_check.rs [hunk 17]
```diff
@@ -1399,26 +1458,26 @@ impl<'a> TypeChecker<'a> {
 									class.phase, class.name, env.phase
 								),
 							);
-							return self.types.error();
+							return (self.types.error(), Phase::Independent);
 						}
 					}
 					// If type is anything we have to assume it's ok to initialize it
-					Type::Anything => return self.types.anything(),
+					Type::Anything => return (self.types.anything(), Phase::Independent),
```
Comment (Chriscbr) [inline libs/wingc/src/type_check.rs:1465]: I wonder if at some point we'll need a `Phase::Unknown` (for now I hope not)

File: libs/wingc/src/type_check.rs [hunk 18]
```diff
@@ -3249,8 +3332,8 @@ impl<'a> TypeChecker<'a> {
 					type_: v,
 					reassignable,
 					phase: flight,
-					is_static,
-					is_member: _,
+					kind,
+					..
```
Comment (Chriscbr) [inline libs/wingc/src/type_check.rs:3336]: Can be removed
```suggestion
					kind,
```

File: examples/tests/invalid/inflight_reassign.w [hunk 19]
```diff
@@ -0,0 +1,10 @@
+let var xvar = "hello";
+let ylet = 123;
+
+inflight () => {
+  xvar = "hi";
+//^^^^ Variable cannot be reassigned from inflight
```
Comment (Chriscbr) [inline examples/tests/invalid/inflight_reassign.w:6]: 👏

**Review by Chriscbr — APPROVED** [review]

File: libs/wingc/src/type_check.rs (hunk 13 above)
Comment (eladb) [reply libs/wingc/src/type_check.rs:80]: It's meant to represent a value that's not bound to an object.

Not sure local is the right term.

We've used "free variables" previously

File: libs/wingc/src/type_check.rs (hunk 16 above)
Comment (eladb) [reply libs/wingc/src/type_check.rs:1326]: OK.

File: libs/wingc/src/type_check.rs (hunk 13 above)
Comment (Chriscbr) [reply libs/wingc/src/type_check.rs:80]: Sorry, I probably didn't explain myself well. I think we're thinking of the term "bound" in different ways (I'm using it in the math/Lisp sense where it refers to a property about substituting variables into functions, and are using it in a way that seems more related to association or membership to types). Anyway I'd be down to chat about it sometime, feels like a shared understanding of compiler terminology is useful long term for taming the complexity monster we have here 😁

**Pull request merged**